curl/src/main/scala/org/http4s/curl/http/CurlRequest.scala

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2022 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.curl.http
+
+import cats.effect._
+import org.http4s.Request
+import org.http4s.Response
+import org.http4s.curl.internal.Utils
+import org.http4s.curl.internal._
+import org.http4s.curl.unsafe.CurlExecutorScheduler
+
+private[curl] object CurlRequest {
+  private def setup(
+      handle: CurlEasy,
+      ec: CurlExecutorScheduler,
+      send: RequestSend,
+      recv: RequestRecv,
+      req: Request[IO],
+  ): Resource[IO, Unit] = Utils.newZone.flatMap(implicit zone =>
+    CurlSList().evalMap(headers =>
+      IO {
+        // TODO add in options
+        // handle.setVerbose(true)
+
+        import org.http4s.curl.unsafe.libcurl_const
+        import scala.scalanative.unsafe._
+        import org.http4s.Header
+        import org.http4s.HttpVersion
+        import org.typelevel.ci._
+
+        handle.setCustomRequest(toCString(req.method.renderString))
+
+        handle.setUpload(true)
+
+        handle.setUrl(toCString(req.uri.renderString))
+
+        val httpVersion = req.httpVersion match {
+          case HttpVersion.`HTTP/1.0` => libcurl_const.CURL_HTTP_VERSION_1_0
+          case HttpVersion.`HTTP/1.1` => libcurl_const.CURL_HTTP_VERSION_1_1
+          case HttpVersion.`HTTP/2` => libcurl_const.CURL_HTTP_VERSION_2
+          case HttpVersion.`HTTP/3` => libcurl_const.CURL_HTTP_VERSION_3
+          case _ => libcurl_const.CURL_HTTP_VERSION_NONE
+        }
+        handle.setHttpVersion(httpVersion)
+
+        req.headers // curl adds these headers automatically, so we explicitly disable them
+          .transform(Header.Raw(ci"Expect", "") :: Header.Raw(ci"Transfer-Encoding", "") :: _)
+          .foreach(header => headers.append(header.toString))
+
+        handle.setHttpHeader(headers.toPtr)
+
+        // NOTE that pointers are to data that is handled by GC
+        // and have proper lifetime management
+        // so no need to handle them explicitly anymore
```

Comment:
Whenever I try tricks like this I feel like it comes back to bite me 😅 where specifically are these references being held?

Reply:
All the references are to objects that have greater lifetime than curl handler transfer.
actually the comment is left from refactoring and is not applicable now anymore. It's from when I was changing `gcRoot.add` stuff.

Reply:
It's a note about
> Lifetime management of objects is passed to GC instead of manual management
this change

Reply:
> All the references are to objects

Yes, where are those references being kept specifically?

Reply:
References are kept in curl handler, and the objects themselves are `RequestSend` and `RequestRecv`, which are managed in the resource that creates a response in `CurlRequest`